File: web/js/lib/job-partition.js
/**
 * Job partitioning + display wording for the Downloads view (WP 4a.5).
 *
 * Splits a `GET /v1/jobs` snapshot into the buckets the Downloads screen
 * renders, against the REAL status set from WP 3.12's job control
 * (api/README.md "Job control" / "The status model") rather than the
 * mockup's fake one. The mockup's statuses are
 * `queued|running|paused|verify|done|error`; the real API has no `verify`
 * — that was a mockup-only transient UI state simulating SteamPrefill's
 * "verifying cached chunks" phase — and adds a real, terminal `cancelled`
 * the mockup never modeled as a distinct outcome (job outcome honesty,
 * docs/PROJECT_PLAN.md: done/error/cancelled render distinctly; a
 * cancelled job is neither a failure nor a success).
 *
 * **The slot-release divergence (docs/WORKPACKAGES.md Phase 4a header,
 * api/README.md "The worker slot — a paused job does NOT hold it").** The
 * mockup assumes exactly one job ever occupies an "active" card slot
 * (running OR paused OR verifying), with the queue waiting behind it. The
 * real backend releases the worker slot on pause: `claim_next_job` claims
 * `queued` rows only, so a paused job's app can sit parked while a
 * completely different queued job is claimed and runs. `running` and
 * `paused` are therefore independent buckets here (normally 0-1 running,
 * 0-N paused), not the mockup's single mutually-exclusive slot — the whole
 * point of the divergence is that the UI can show "Active" (what is really
 * using the one worker slot right now) and "Paused" (parked, NOT blocking
 * anything) as genuinely separate sections instead of one that pretends a
 * paused job still occupies a slot it does not hold.
 *
 * **Unknown status routes to history, not oblivion (WP 4a.8 backport of the
 * WP 4b.5 divergence, docs/WORKPACKAGES.md's Phase 4a header).** A job whose
 * `status` is anything other than the six known values used to match none
 * of this module's filters and silently vanish from every bucket — an
 * operator-invisible job that still exists server-side. `partitionJobs` now
 * routes it into `history` with a NEUTRAL presentation instead:
 * `jobIconKind` already fell back to `"none"` for an unmapped status and
 * `jobStatusWord` already fell back to the raw status string rather than
 * fabricating a plausible-looking word, so nothing else needed to change to
 * make that an honest "something happened, here it is, we don't have a name
 * for it" row. It deliberately does NOT count toward `countPending`: an
 * unrecognized status is not confidently "still pending" either, so the nav
 * pip stays fail-quiet rather than fail-loud on every poll tick. Ported
 * from the Android sibling's `ui/downloads/logic/JobPartition.kt`
 * (`KNOWN_STATUSES`/history-filter shape) verbatim — ONE change from that
 * port: the Kotlin side stayed 4a.5-shaped ("earlier port had no name for
 * this"), this backport closes the same gap here.
 *
 * Pure only — no DOM, no fetch. Covered in web/tests/job-partition.test.js.
 */

/** Statuses that count toward "something is pending" (the Downloads nav
 * pip, mockup's `syncPip`) — ported onto the real status set (no
 * `verify`). Mirrors `JOB_ACTIVE_STATUSES` in store.js/library.js, kept as
 * an independent local copy rather than a shared import: this module must
 * stay standalone (no view-layer coupling), and the two lists are pinned
 * to the same value by api/README.md's "ACTIVE_STATUSES" table, not by
 * each other. */
const PENDING_STATUSES = new Set(["queued", "running", "paused"]);
const HISTORY_STATUSES = new Set(["done", "error", "cancelled"]);
const KNOWN_STATUSES = new Set(["queued", "running", "paused", "done", "error", "cancelled"]);

/**
 * @param {object[] | null | undefined} jobs `GET /v1/jobs` snapshot.
 * @returns {{running: object[], paused: object[], queued: object[], history: object[]}}
 *   `queued` is sorted oldest-job-id-first (FIFO draw order,
 *   api/README.md "Queue semantics" — "exactly one job runs at a time...
 *   FIFO by job id"), independent of the snapshot's own order. `history`
 *   keeps the snapshot's order as-is (`GET /v1/jobs` is newest-first per
 *   api/README.md, so a job's most recent occurrence is already first) —
 *   an unrecognized status is included here too (module header, "Unknown
 *   status routes to history"), appended wherever the snapshot placed it.
 */
export function partitionJobs(jobs) {
  const list = Array.isArray(jobs) ? jobs : [];
  const running = list.filter((j) => j.status === "running");
  const paused = list.filter((j) => j.status === "paused");
  const queued = list.filter((j) => j.status === "queued").sort((a, b) => a.id - b.id);
  const history = list.filter((j) => HISTORY_STATUSES.has(j.status) || !KNOWN_STATUSES.has(j.status));
  return { running, paused, queued, history };
}

/** How many jobs the Downloads nav pip should count (mockup's `syncPip`,
 * ported onto the real status set). */
export function countPending(jobs) {
  const list = Array.isArray(jobs) ? jobs : [];
  return list.filter((j) => PENDING_STATUSES.has(j.status)).length;
}

/**
 * 1-based queue position of `jobId` within an already-FIFO-sorted `queued`
 * array (as returned by `partitionJobs`). `null` if not present.
 */
export function queuePosition(queued, jobId) {
  const list = Array.isArray(queued) ? queued : [];
  const idx = list.findIndex((j) => j.id === jobId);
  return idx === -1 ? null : idx + 1;
}

/** Which status-icon KIND (js/components/status-icon.js) a job's badge
 * uses. `queued` has no persistent badge in this view (mockup parity: the
 * queue row is grip + name + position, no icon) — `"none"` is returned for
 * completeness/testability, not for display. */
export function jobIconKind(job) {
  switch (job.status) {
    case "running":
      return "running";
    case "paused":
      return "paused";
    case "done":
      return "cached";
    case "error":
      return "error";
    case "cancelled":
      return "cancelled";
    default:
      return "none";
  }
}

// GC jobs are real `GET /v1/jobs` rows (type: "gc") the mockup's JOBS
// fixture never modeled (NOTES.md has no GC entries) — "Downloading" /
// "Update ready" wording is honest for a prefill job and misleading for a
// garbage-collection one, so GC gets its own words while sharing the same
// status-icon KIND per jobIconKind (no dedicated GC glyph exists yet;
// documented simplification, not an oversight — WP 4a.4 owns the GC
// trigger flow, this view only ever DISPLAYS a GC job it happens to see).
const PREFILL_WORD = Object.freeze({
  queued: "Queued",
  running: "Downloading",
  paused: "Paused",
  done: "Done",
  error: "Failed",
  cancelled: "Cancelled",
});
const GC_WORD = Object.freeze({
  queued: "Garbage collection queued",
  running: "Collecting garbage",
  paused: "Paused", // unreachable in practice (pause 409s on a GC job) — defensive only
  done: "Garbage collected",
  error: "Garbage collection failed",
  cancelled: "Garbage collection cancelled",
});

/**
 * The word a job's badge/history-row shows, single source of truth for
 * both (mockup's `LABEL` table split by job type). Falls back to the raw
 * status string for a status this table doesn't know, rather than
 * fabricating something plausible-looking.
 * @param {{type: string, status: string}} job
 */
export function jobStatusWord(job) {
  const table = job.type === "gc" ? GC_WORD : PREFILL_WORD;
  return table[job.status] || job.status;
}

File: app/app/src/main/java/dev/steamvault/app/ui/downloads/logic/JobPartition.kt
package dev.steamvault.app.ui.downloads.logic

import dev.steamvault.app.net.model.JobSummary

/**
 * Job partitioning + display wording for the Downloads screen (WP 4b.5) —
 * Kotlin port of `web/js/lib/job-partition.js` onto the same real WP 3.12
 * status set (api/README.md "Job control" / "The status model" / "The
 * worker slot").
 *
 * **The slot-release divergence, ported unchanged (api/README.md "The
 * worker slot — a paused job does NOT hold it", docs/WORKPACKAGES.md Phase
 * 4a header).** The frozen mockup assumes one job ever occupies a single
 * "active" slot (running OR paused), with the queue waiting behind it. The
 * real backend releases the worker slot on pause: `claim_next_job` claims
 * `queued` rows only, so a paused job's app can sit parked while a
 * completely different queued job is claimed and runs. [partitionJobs]
 * therefore treats `running` and `paused` as INDEPENDENT buckets (normally
 * 0-1 running, 0-N paused), not one mutually-exclusive slot — this is what
 * lets the Downloads screen render "Active" (what really holds the one
 * worker slot right now) and "Paused" (parked, holding nothing) as genuinely
 * separate sections instead of one that pretends a paused job still
 * occupies a slot it does not hold. Web's own screen states this note on
 * the paused card itself; this app's screen carries the identical wording
 * (see `ui/downloads/DownloadsStrings.kt` / `strings.xml`'s
 * `downloads_paused_hold_note`, `downloads_queue_hint_*`).
 *
 * **Divergence from the web port — unknown status is NOT silently dropped
 * (WP 4a.5 review nit, recorded in docs/WORKPACKAGES.md's Phase 4a
 * header).** `web/js/lib/job-partition.js`'s `partitionJobs` buckets a job
 * only when its status is exactly one of
 * running/paused/queued/done/error/cancelled — a row whose status is
 * anything else (a future status this client doesn't know about yet, or a
 * corrupted value) matches none of the web module's filters and
 * silently vanishes from every bucket: an operator-invisible job that still
 * exists server-side. This port instead routes an unrecognized status into
 * [PartitionedJobs.history] with a NEUTRAL presentation —
 * [jobIconKind] already falls back to `"none"` for an unmapped status
 * (mirrors [dev.steamvault.app.ui.status.StatusKind.fromWireName]'s own
 * unknown-kind fallback) and [jobStatusWord] already falls back to the raw
 * status string rather than fabricating a plausible-looking word — so an
 * honest "something happened, here it is, we don't have a name for it"
 * beats a job that disappears from the screen entirely. It deliberately
 * does NOT count toward [countPending]: an unrecognized status is not
 * confidently "still pending" either, so the nav pip stays fail-quiet
 * rather than fail-loud on the one signal that would page an operator for
 * every poll tick. This is an IMPROVEMENT over the web behaviour, not a
 * behavioural requirement carried over from it — recorded here rather than
 * silently, per the WP brief.
 *
 * Pure only — no Compose, no coroutines. Covered by `JobPartitionTest`.
 */
data class PartitionedJobs(
    val running: List<JobSummary>,
    val paused: List<JobSummary>,
    /** FIFO draw order (oldest job id first, api/README.md "Queue
     * semantics"), independent of the snapshot's own order. */
    val queued: List<JobSummary>,
    /** Keeps the snapshot's own order (`GET /v1/jobs` is newest-first per
     * api/README.md, so a job's most recent occurrence is already first). */
    val history: List<JobSummary>,
)

/** Statuses that count toward "something is pending" (the Downloads nav
 * pip) — mirrors [dev.steamvault.app.polling.PollingIntervals]'s own
 * `JOB_ACTIVE_STATUSES`, kept as an independent local copy rather than a
 * shared import: this module must stay standalone, and the two sets are
 * pinned to the same value by api/README.md's status table, not by each
 * other (same reasoning `web/js/lib/job-partition.js`'s header gives for
 * its own `PENDING_STATUSES` copy). */
private val PENDING_STATUSES = setOf("queued", "running", "paused")
private val HISTORY_STATUSES = setOf("done", "error", "cancelled")
private val KNOWN_STATUSES = setOf("queued", "running", "paused", "done", "error", "cancelled")

/**
 * @param jobs `GET /v1/jobs` snapshot.
 */
fun partitionJobs(jobs: List<JobSummary>): PartitionedJobs {
    val running = jobs.filter { it.status == "running" }
    val paused = jobs.filter { it.status == "paused" }
    val queued = jobs.filter { it.status == "queued" }.sortedBy { it.id }
    // Improvement over the web port (see module kdoc): an unrecognized
    // status lands in history too, instead of matching no filter at all.
    val history = jobs.filter { it.status in HISTORY_STATUSES || it.status !in KNOWN_STATUSES }
    return PartitionedJobs(running, paused, queued, history)
}

/** How many jobs the Downloads nav pip should count. */
fun countPending(jobs: List<JobSummary>): Int = jobs.count { it.status in PENDING_STATUSES }

/**
 * 1-based queue position of [jobId] within an already-FIFO-sorted `queued`
 * list (as returned by [partitionJobs]). `null` if not present.
 */
fun queuePosition(queued: List<JobSummary>, jobId: Int): Int? {
    val idx = queued.indexOfFirst { it.id == jobId }
    return if (idx == -1) null else idx + 1
}

/** Which [dev.steamvault.app.ui.status.StatusKind] wire name a job's badge
 * uses. `queued` has no persistent badge in this view (mockup parity: the
 * queue row is grip + name + position, no icon) — `"none"` is returned for
 * completeness/testability, not for display. An unrecognized status also
 * falls back to `"none"`, same fallback [dev.steamvault.app.ui.status.StatusKind.fromWireName]
 * itself already applies. */
fun jobIconKind(job: JobSummary): String = when (job.status) {
    "running" -> "running"
    "paused" -> "paused"
    "done" -> "cached"
    "error" -> "error"
    "cancelled" -> "cancelled"
    else -> "none"
}

// GC jobs are real `GET /v1/jobs` rows (type: "gc") — "Downloading" /
// "Update ready" wording is honest for a prefill job and misleading for a
// garbage-collection one, so GC gets its own words while sharing the same
// status-icon kind via jobIconKind (mirrors web/js/lib/job-partition.js).
private val PREFILL_WORD = mapOf(
    "queued" to "Queued",
    "running" to "Downloading",
    "paused" to "Paused",
    "done" to "Done",
    "error" to "Failed",
    "cancelled" to "Cancelled",
)
private val GC_WORD = mapOf(
    "queued" to "Garbage collection queued",
    "running" to "Collecting garbage",
    "paused" to "Paused", // unreachable in practice (pause 409s on a GC job) — defensive only
    "done" to "Garbage collected",
    "error" to "Garbage collection failed",
    "cancelled" to "Garbage collection cancelled",
)

/**
 * The word a job's badge/history-row shows, single source of truth for
 * both (mirrors web's `LABEL` table split by job type). Falls back to the
 * raw status string for a status this table doesn't know, rather than
 * fabricating something plausible-looking (same posture [partitionJobs]'s
 * unknown-status handling takes).
 */
fun jobStatusWord(job: JobSummary): String {
    val table = if (job.type == "gc") GC_WORD else PREFILL_WORD
    return table[job.status] ?: job.status
}
